Readme updates (#2095)
rancher/k3s: README.md

What is this?
---

K3s is a [fully conformant](https://github.com/cncf/k8s-conformance/pulls?q=is%3Apr+k3s) Kubernetes distribution with the following changes:

1. It is packaged as a single binary.
1. It swaps out etcd for a lightweight sqlite3 storage backend. etcd3, MySQL, and Postgres also supported.
1. It wraps Kubernetes and other components in a single, simple launcher.
1. It is secure by default with reasonable defaults for lightweight environments.
1. It has minimal to no OS dependencies (just a sane kernel and cgroup mounts needed).

> k3s bundles the following technologies together into a single cohesive distibution:

Reviewer: Should "K3s" be consistent throughout the doc in the context of a proper name? Seems there's variance in the capitalization.

Author: i'm going to go with "capital only at the beginning of sentences" for now. In which case, i see now that I do have one mistake. Other places follow this convention

What is this?
---

K3s is a [fully conformant](https://github.com/cncf/k8s-conformance/pulls?q=is%3Apr+k3s) Kubernetes distribution with the following changes:

1. It is packaged as a single binary.

> 1. It swaps out etcd for a lightweight sqlite3 storage backend. etcd3, MySQL, and Postgres are also supported.

Reviewer 2: I'd say "add support for..... in addition to etcd."

1. It wraps Kubernetes and other components in a single, simple launcher.
1. It is secure by default with reasonable defaults for lightweight environments.

> 1. It has minimal to no OS dependencies (just a sane kernel and cgroup mounts needed).

Reviewer 3: Maybe also call out the simplified server-agent connectivity? Requires only a single open port on servers, no ports on agents (not counting whatever you need for CNI of course).

K3s bundles the following technologies together into a single cohesive distibution:

* [Containerd](https://containerd.io/) & [runc](https://github.com/opencontainers/runc)
* [Flannel](https://github.com/coreos/flannel) for CNI
* [CoreDNS](https://coredns.io/)

> * [Metric-server](https://github.com/kubernetes-sigs/metrics-server)

Reviewer 3: Plural, and is styled in the upstream docs with a space in the name:
```suggestion
* [Metrics Server](https://github.com/kubernetes-sigs/metrics-server)
```

* [Traefik](https://containo.us/traefik/) for ingress
* [Klipper-lb](https://github.com/rancher/klipper-lb) as an embedded service loadbalancer provider
* [Kube-router](https://www.kube-router.io/) for network policy
* [Helm-controller](https://github.com/rancher/helm-controller) to allow for CRD-driven deployment of helm manifests
* [Kine](https://github.com/rancher/kine) as a datastore shim that allows etcd to be replaced with other databases
* [Local-path-provisioner](https://github.com/rancher/local-path-provisioner) for provisioning volumes using local storage
* [Host utilities](https://github.com/rancher/k3s-root) such as iptables/nftables, ebtables, ethtool, & socat

Additionally, K3s simplifies Kubernetes operations by maintaining functionality for:

* Managing the TLS certificates of Kubernetes componenents
* Managing the connection between worker and server nodes
* Managing an embedded etcd cluster

> * Auto-deploying Kubernetes resources on startup

Reviewer 3: I can't think of wording that I love, but it might be worth calling out that it doesn't just deploy on startup - it's constantly deploying the files as they are changed.
```suggestion
* Auto-deploying Kubernetes resources from local manifests, in realtime as they are changed.
```

What's with the name?
--------------------
We wanted an installation of Kubernetes that was half the size in terms of memory footprint. Kubernetes is a
10 letter word stylized as k8s. So something half as big as Kubernetes would be a 5 letter word stylized as
K3s. There is no long form of K3s and no official pronunciation.

> Is this a fork?

Reviewer 2: I'd also call out that k3s is a distribution.  Once a user starts k3s they should have a fully functioning cluster with all major APIs support.  This is what differentiates say kubeadm from k3s. Kubeadm does not include additional components like ingress, storage classs, network policy, service lb.

Reviewer 2: And those extra components essentially differential a distro from k8s upstream.

---------------

> No. A fork implies continued divergence from the original. This is not K3s's goal or practice. K3s seeks to remain as close to upstream Kubernetes as possible. We do maintain a small set of patches (well under 1000 lines) important to K3s's usecase and deployment model. We maintain patches for other components as well. When possible, we contribute these changes back to the upstream projects, for example with [SELinux support in containerd](https://github.com/containerd/cri/pull/1487/commits/24209b91bf361e131478d15cfea1ab05694dc3eb). This is a common practice amongst software distributions.

Reviewer 2: I'd call out that it is not the intention of k3s to change kubernetes to add or change behavior of its core features.  The changes are primarily done to assist how we package. I know you already said half of that.

Release cadence
-------------------
K3s maintains pace with upstream Kubernetes releases. Our goal is to release patch releases on the same day as upstream and minor releases within a few days.

Our release versioning reflects the version of upstream Kubernetes that is being released. For example, the K3s release [v1.18.6+k3s1](https://github.com/rancher/k3s/releases/tag/v1.18.6%2Bk3s1) maps to the `v1.18.6` Kubernetes release. We add a postfix in the form of `+k3s<number>` to allow us to make additional releases using the same version of upstream Kubernetes, while remaining [semver](https://semver.org/) compliant. For example, if we discovered a high severity bug in `v1.18.6+k3s1` and needed to release an immediate fix for it, we would release `v1.18.6+k3s2`.

Documentation
-------------

Please see [the official docs site](https://rancher.com/docs/k3s/latest/en/) for complete documentation.

Quick-Start - Install Script
--------------

> The `install.sh` script provides a convenient way for installing to systemd or openrc,

Reviewer 3: ```suggestion
The `install.sh` script provides a convenient way to download K3s and add a service to systemd or openrc - 
```

* Managing an embedded etcd cluster
* Auto-deploying Kubernetes resources from local manifests, in realtime as they are changed.

What's with the name?
--------------------
We wanted an installation of Kubernetes that was half the size in terms of memory footprint. Kubernetes is a
10 letter word stylized as k8s. So something half as big as Kubernetes would be a 5 letter word stylized as
K3s. There is no long form of K3s and no official pronunciation.

Is this a fork?
---------------
No, it's a distribution. A fork implies continued divergence from the original. This is not K3s's goal or practice. K3s explicitly intends to not change any core Kubernetes functionality. We seek to remain as close to upstream Kubernetes as possible. We do maintain a small set of patches (well under 1000 lines) important to K3s's usecase and deployment model. We maintain patches for other components as well. When possible, we contribute these changes back to the upstream projects, for example with [SELinux support in containerd](https://github.com/containerd/cri/pull/1487/commits/24209b91bf361e131478d15cfea1ab05694dc3eb). This is a common practice amongst software distributions.

> K3s is a distribution because it packages additional components and services necessary for a fully functional cluster that go beyond vanilla Kubernetes. These are opinionated choices on technologies for components like ingress, storage class, network policy, service load balancer, and even container runtime. These choices and technologies are touched on in more detail in the [What is this?](what-is-this) section.

Author: Addressed this by adding a line to the "What is this?" as opposed to down here.
https://github.com/rancher/k3s/pull/2095/commits/0a02996a6f229ed67daa2b0cb50e2a02a9c8a1fc

How is this lightweight or smaller than upstream Kubernetes?
---
There are two major ways that K3s is lighter weight than upstream Kubernetes:
1. The memory footprint to run it is smaller
2. The binary, which contains all the non-containerized components needed to run a cluster, is smaller

The memory footprint is reduced primarily by running many components inside of single process. This eliminates significant overhead that would otherwise be duplicated for each component.

The binary is smaller primarily by removing some unused Kubernetes code.

What have you removed from upstream Kubernetes?
---
This is a common point of confusion because it has changed over time. Early versions of K3s had much more removed than current version. Current versions of K3s remove two things:
1. In-tree storage drivers
2. In-tree cloud provider

> We remove these to achieve a smaller binary size. They can be removed while remaining conformant because neither affect core Kubernetes functionality. They are also dependent on cloud or data center technologies/services, which are generally not available in K3s's usecases. Finally, both have out-of-tree alternatives in the form of [CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md) and [CCM](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/), which work in K3s and which upstream is moving towards.

Reviewer 4: The crucial point about tailored support through these out-of-tree components feesl like it could get lost in this paragraph.

The memory footprint is reduced primarily by running many components inside of single process. This eliminates significant overhead that would otherwise be duplicated for each component.

The binary is smaller by disabling third-party storage drivers and cloud providers, which is explained in more detail below.

What have you removed from upstream Kubernetes?
---

> This is a common point of confusion because it has changed over time. Early versions of K3s had much more removed than current version. Current versions of K3s disables two things:

Reviewer 3: "current versions of k3s disable" or "k3s currently disables"
